libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewRouteEngineTest.java

```diff
@@ -0,0 +1,184 @@
+import static org.mockito.Mockito.verify;
+
+public class NavigationViewRouteEngineTest extends BaseTest {
+
+  private static final String DIRECTIONS_PRECISION_6 = "directions_v5_precision_6.json";
+
+  @Test
+  public void sanity() throws Exception {
+    NavigationViewRouteEngineListener routeEngineListener = mock(NavigationViewRouteEngineListener.class);
+    NavigationViewRouteEngine routeEngine = buildRouteEngine(routeEngineListener);
+
+    assertNotNull(routeEngine);
+  }
+
+  @Test
+  public void onExtractOptionsWithRoute_routeUpdateCallbackIsCalled() throws Exception {
+    NavigationViewRouteEngineListener routeEngineListener = mock(NavigationViewRouteEngineListener.class);
+    NavigationViewRouteEngine routeEngine = buildRouteEngine(routeEngineListener);
+    NavigationViewOptions options = buildNavigationViewOptionsWithRoute();
+    Context mockContext = mock(Context.class);
+    DirectionsRoute directionsRoute = options.directionsRoute();
+
+    routeEngine.extractRouteOptions(mockContext, options);
+
+    verify(routeEngineListener).onRouteUpdate(directionsRoute);
+  }
+
+  @Test
+  public void onExtractOptionsWithRoute_destinationCallbackIsCalled() throws Exception {
+    NavigationViewRouteEngineListener routeEngineListener = mock(NavigationViewRouteEngineListener.class);
+    NavigationViewRouteEngine routeEngine = buildRouteEngine(routeEngineListener);
+    NavigationViewOptions options = buildNavigationViewOptionsWithRoute();
+    Point destination = findDestinationPoint(options);
+    Context mockContext = mock(Context.class);
+
+    routeEngine.extractRouteOptions(mockContext, options);
+
+    verify(routeEngineListener).onDestinationSet(destination);
+  }
+
+  @Test
+  public void onExtractOptionsWithCoordinates_destinationCallbackIsCalled() throws Exception {
+    NavigationViewRouteEngineListener routeEngineListener = mock(NavigationViewRouteEngineListener.class);
+    NavigationViewRouteEngine routeEngine = buildRouteEngine(routeEngineListener);
+    NavigationViewOptions options = buildNavigationViewOptionsWithCoordinates();
+    Context mockContext = mock(Context.class);
+    Point destination = options.destination();
+
+    routeEngine.extractRouteOptions(mockContext, options);
+
+    verify(routeEngineListener).onDestinationSet(destination);
+  }
+
+  @Test
+  public void onResponseReceived_routeIsProcessed() throws Exception {
+    NavigationViewRouteEngineListener routeEngineListener = mock(NavigationViewRouteEngineListener.class);
+    NavigationViewRouteEngine routeEngine = buildRouteEngine(routeEngineListener);
+    DirectionsResponse response = buildDirectionsResponse();
+    DirectionsRoute route = response.routes().get(0);
+    RouteProgress routeProgress = mock(RouteProgress.class);
+
+    routeEngine.onResponseReceived(response, routeProgress);
+
+    verify(routeEngineListener).onRouteUpdate(route);
+  }
+
+  @Test
+  public void onSecondResponseReceived_routesAreCompared() throws Exception {
```

Comment:
We should be explicit on `onSecondResponseReceived` somehow to know what's going on here in a glance. Are we enabling some option in `buildNavigationEngineWithOptions`? Wondering how this test differs from `onResponseReceived_routeIsProcessed` one 🤔